test/test_filepost.py

```diff
@@ -131,3 +168,195 @@ def test_request_fields(self):
           b'v\r\n'
           b'--' + b(BOUNDARY) + b'--\r\n'
           )
+
+
+class _SeekableObject(object):
+    def __init__(self, data):
+        self.data = io.BytesIO(data)
+
+    def seek(self, *args, **kwargs):
+        return self.data.seek(*args, **kwargs)
+
+    def tell(self, *args, **kwargs):
+        return self.data.tell(*args, **kwargs)
+
+
+class _ReadableObject(object):
+    def __init__(self, data):
+        self.data = io.BytesIO(data)
+
+    def read(self, chunk=None):
+        return self.data.read(chunk)
+
+
+class _IterableObject(object):
+    def __init__(self, data):
+        self.data = io.BytesIO(data)
+
+    def __iter__(self):
+        return iter(self.data)
+
+
+class _LengthObject(object):
+    def __init__(self, data):
+        self.length = len(data)
+
+    def __len__(self):
+        return self.length
+
+
+class TestStreamingUploads(unittest.TestCase):
+    def test_length_file_like_object(self):
+        fields = [('k', ('somefile.txt', io.BytesIO(b'v'), 'image/jpeg'))]
+
+        encoded, _ = encode_multipart_formdata(fields, boundary=BOUNDARY)
+        self.assertEqual(len(encoded), 146)
+
+    def test_length_readable_object(self):
+        fields = [('k', ('somefile.txt', _ReadableObject(b'v'), 'image/jpeg'))]
+
+        encoded, _ = encode_multipart_formdata(fields, boundary=BOUNDARY)
+        self.assertEqual(len(encoded), 146)
+
+    def test_length_iterable_object(self):
+        fields = [('k', ('somefile.txt', _IterableObject(b'v'), 'image/jpeg'))]
+
+        encoded, _ = encode_multipart_formdata(fields, boundary=BOUNDARY)
+        self.assertEqual(len(encoded), 146)
+
+    def test_length_unknown_object(self):
+        fields = [('k', ('somefile.txt', object(), 'image/jpeg'))]
+
+        encoded, _ = encode_multipart_formdata(fields, boundary=BOUNDARY)
+        self.assertRaises(TypeError, encoded.__len__)
+
+    def test_unknown_object(self):
+        unknown_object = object()
+        fields = [('k', ('somefile.txt', unknown_object, 'image/jpeg'))]
+
+        encoded, _ = encode_multipart_formdata(fields, boundary=BOUNDARY, chunk_size=1)
+        non_chunks = [b'--' + b(BOUNDARY) + b'\r\n',
+                      b'Content-Disposition: form-data; name="k"; filename="somefile.txt"\r\n'
+                      b'Content-Type: image/jpeg\r\n\r\n',
+                      b'\r\n',
+                      b'--' + b(BOUNDARY) + b'--\r\n']
+
+        for chunk in encoded:
+            if chunk not in non_chunks:
+                self.assertIs(chunk, unknown_object)
```

Comment:
I don't know that this test is well-crafted. It might be better to use equality the whole way through and place the unknown object in the right place.